Mathlib/RingTheory/Kaehler.lean

```diff
@@ -388,18 +385,42 @@ theorem KaehlerDifferential.End_equiv_aux (f : S →ₐ[R] S ⊗ S ⧸ KaehlerDi
     exact e₁.symm.trans (e.trans e₂)
 #align kaehler_differential.End_equiv_aux KaehlerDifferential.End_equiv_aux
 
-set_option maxHeartbeats 700000 in
--- Porting note: extra heartbeats are needed to infer the instance
--- Module S { x // x ∈ Ideal.cotangentIdeal (ideal R S) }
-set_option synthInstance.maxHeartbeats 200000 in
--- This has type
--- `Derivation R S (Ω[S⁄R]) ≃ₗ[R] Derivation R S (KaehlerDifferential.ideal R S).cotangentIdeal`
--- But lean times-out if this is given explicitly.
+/- After porting note: Lean is slow to synthesize theses instances.
+  Without them the endEquivDerivation' and endEquivAuxEquiv both have
+  significant timeouts. This was not the case in Mathlib 3 -/
+local instance : SMul (S ⊗[R] S ⧸ KaehlerDifferential.ideal R S ^ 2)
+    (S ⊗[R] S ⧸ KaehlerDifferential.ideal R S ^ 2) := Algebra.toSMul
```

Comment:
`Algebra.toSMul` is not supposed to ever be used (it has priority 0)

Reply:
https://github.com/leanprover-community/mathlib4/blob/86ac6bf9dd506d502b43d848103031eaf1e8bc7c/Mathlib/Algebra/Algebra/Basic.lean#L344-L345

Reply:
I don't understand why we deleted that, I think it should be restored (and I'd be interested to know how that affects performance)

Reply:
I can restore it but the answer is that adds extra tc synthesis passes on goals that are already slow (see above).

Reply:
> `Algebra.toSMul` is not supposed to ever be used (it has priority 0)

That’s what it finds naturally. I’m sure `Mul.toSMul` will be fine.

Reply:
My understanding is that performance-wise, `Algebra.toSMul` is a bad instance because it shortcuts `Algebra.toModule, `Module.toDistribMulAction`, `DistribMulAction.toMulAction`, and `MulAction.toSMul`. We shouldn't have transitive shortcuts, because then the number of instances we have grows massively.

I think @digama0 has expressed this opinion before?

Reply:
That makes sense. `Mul.toSMul` keeps things below the timeout but seems (to the human eye) to elaborate more slowly

Reply:
De-instanced `Algebra.toSMul` in #6219. Experiments indicate it would slow things down even more.

Reply:
> I think @digama0 has expressed this opinion before?

I have expressed the opposite, basically lamenting that we can't have these instances because they would significantly decrease the cost of adding new intermediate classes if we could have shorter "expressways" which skip the intermediates and are preferred. Having them greatly increases the (already exponential) number of paths through the graph, but whether this is a crippling problem or not at all hinges rather delicately on details of the typeclass inference algorithm and we have gone back and forth between preferring and dispreferring transitive shortcut instances.